rfcs/0036-rfc-process-team-amendment.md

```diff
@@ -0,0 +1,202 @@
+   in the comment thread of the pull request itself. Discussion outside of the
+   pull request, either offline or in a video conference, that might be
+   preferable to get to a solution for complex issues, will be summarized on the
+   pull request comment thread.
+8. RFCs rarely go through this process unchanged, especially as alternatives and
+   drawbacks are shown. You can make edits, big and small, to the RFC to clarify
+   or change the design, but make changes as new commits to the pull request,
+   and leave a comment on the pull request explaining your changes.
+   Specifically, do not squash or rebase commits after they are visible on the
+   pull request.
+9. At some point, a member of the RFC Shepherd Team will propose a "motion for
+final comment period" (FCP), along with a disposition for the RFC (merge, close,
+or postpone).
+    * This step is taken when enough of the tradeoffs have been discussed that
+      the RFC Shepherd Team is in a position to make a decision. That does not
+      require consensus amongst all participants in the RFC thread (which is
+      usually impossible). However, the argument supporting the disposition on
+      the RFC needs to have already been clearly articulated, and there should
+      not be a strong consensus against that position outside of the RFC
+      Shepherd Team. RFC Shepherd Team members use their best judgment in taking
+      this step, and the FCP itself ensures there is ample time and notification
+      for stakeholders to push back if it is made prematurely.
+    * For RFCs with lengthy discussion, the motion to FCP is usually preceded by
+      a summary comment trying to lay out the current state of the discussion
+      and major tradeoffs/points of disagreement.
+    * Before actually entering FCP, all members of the RFC Shepherd Team must
+      sign off the motion.
+10. The FCP lasts ten calendar days, so that it is open for at least 5 business
+days. It is also advertised widely, e.g. in NixOS Weekly and through Discourse
+announcements. This way all stakeholders have a chance to lodge any final
+objections before a decision is reached.
+11. In most cases, the FCP period is quiet, and the RFC is either merged or
+closed. However, sometimes substantial new arguments or ideas are raised, the
+FCP is canceled, and the RFC goes back into development mode.
+12. In case of acceptance, the RFC Steering Committee merges the PR into the
+`accepted`, in case of rejection into the `rejected` directory.
+
+
+![RFC Process](./0036-rfc-process.png)
+![Review Process](./0036-review-process.png)
+
+
+## The RFC life-cycle
+
+Once an RFC is accepted the authors may implement it and submit the feature as a
+pull request to the Nix or nixpkgs repo. Being accepted is not a rubber stamp,
+and in particular still does not mean the feature will ultimately be merged; it
+does mean that in principle all the major stakeholders have agreed to the
+feature and are amenable to merging it.
```

Comment:
Should we explicitly state that a positive RFC switches the default to "no reply means no go" to "no reply means go ahead?" i.e. if some feature goes through the RFC process, then the implementation is PR'd and sits for a reasonable amount of time with the appropriate owners pinged but no substantive negative feedback and otherwise stalled conversation, should anyone with the appropriate commit bit feel free to merge?

Reply:
I think if we expect such a change of perception, it needs to be explicit in the RFC. If this amendment gets direct objections, at least we will know what is the situation…

Reply:
I'd expect the implementer to @mention the corresponding shepherd team, and as a result we should have at least one review of the implementation.  RFCs are supposed to be about more significant changes, and I wouldn't like to merge those without at least a second pair of eyes on the code changes.

Yes, there might be some exceptions, but as today &ndash; it ultimately depends on the judgement of the one who pushes/merges a change without any reaction.  For those "trivial" RFC implementations, it should be easy enough to get another opinion like _"I haven't really looked, but this kind of change seems OK without review"._

Reply:
> it ultimately depends on the judgement of the one who pushes/merges a change without any reaction.

I think our community has shown time and time again that that judgment will almost always err in favor of inaction. We're a respectful and cautious bunch, if we don't explicitly say somewhere that it's OK to go ahead in most cases we won't. If that's the desired outcome that's one thing, but if the expectation is that an accepted RFC will generally mean that a sound PR will get merged, then we need to be clear about that.

Reply:
> [...] the expectation is that an accepted RFC will generally mean that a sound PR will get merged [...]

I agree with the expectation, though it's hard to say how strong it should be.  I'd say it might be individual for each RFC, judged from the discussion around it.  Sometimes it's difficult to anticipate all important information _before_ making an implementation and yet you may want an RFC to gather some more formal opinion before undertaking full implementation.

The current text does express something like what you write, though I expect you'd prefer a stronger wording:

> Being accepted is not a rubber stamp, and in particular still does not mean the feature will ultimately be merged; it does mean that in principle all the major stakeholders have agreed to the feature and are amenable to merging it.

Reply:
I would say that given the project's track record on reaching a decision or at least finding out the reason why decision has not been reached, this wording doesn't really promise that RFCs mean anything.

I would have more hopes for a text like 

> Of course, an implementation PR for an accepted RFC can still be rejected due to implementation-specific concerns, or due to some substantial issues not discussed in the context of RFC that become apparent in the context of an implementation.

I guess that would be too optimistic:

> If substantial new (or implementation-specific) issues are not raised in a reasonable time, it would be acceptable to merge for any willing person with commit access after a single positive detailed implementation review (by a person who didn't participate in the implementation).

Reply:
Maybe there should be a deadline for review in the implementation PR, just like there is for the RFC PR? The sheperds could be responsible for this: say a implementation PR musst enger final review after x days of no new comments, and then must either be closed or merged after 10 days. This should prevent stalled PRs at least.

Reply:
We haven't added that on purpose, since the discussion should continue without time limit as long as it is constructive, new points are brought up and the RFC is iterated on. If this is the case no longer, then the Shepherd Team shall step in with a motion for FCP.